### DISCUSSION TITLE
Extend the UI with plugin

### DISCUSSION THREAD

**Discussion by YeomanYe** [eclipse-theia/theia#10975] [Q&A] [answered]
Can I extend theia IDE's UI via plugin, for example: I want to create a custom sidebar and use my own UI to create a graphical description like the AST syntax tree. What should I do if I can.

![image](https://user-images.githubusercontent.com/18027354/161264771-1613fa7b-8afe-4639-88e3-14c2969dee21.png)

ast graph

![image](https://user-images.githubusercontent.com/18027354/161264523-63f92dc6-19c5-4ba7-b0dc-1f22043d2135.png)

**Comment by msujew** [comment, marked as answer]
Hi @YeomanYe,

While it's probably possible to composite a new widget based on an existing editor widget and some custom stuff, I wouldn't recommend going that way. Aside from the yet unknown issues which might pop up in the editor, there's a more VSCode/Theia-like way of accomplishing what you want. 

I would recommend to write a VSCode extension which adds a toolbar item to the `.wf` file extension and opens a preview-like diagram. We have written something similar to integrate a DSL with the visualization framework Sprotty in VSCode. See [this repository](https://github.com/eclipse/sprotty-vscode) for some images of how this would look like and to get some idea how one would accomplish this using the VSCode extension API.

You can then simply install the extension in Theia, if you want so.
